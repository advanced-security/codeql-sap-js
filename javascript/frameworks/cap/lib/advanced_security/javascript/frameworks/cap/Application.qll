/**
 * Definitions pertaining to the application as a whole.
 */

import javascript
import advanced_security.javascript.frameworks.cap.PackageJson

class RootDirectory extends Folder {
  RootDirectory() {
    exists(PackageJson packageJson | this = packageJson.getJsonFile().getParentContainer())
  }

  /**
   * Gets the path of a file relative to this root directory.
   */
  string getFilePathRelativeToRoot(File file) {
    result = file.getAbsolutePath().regexpReplaceAll(this.getAbsolutePath(), ".") and
    result.charAt(0) = "."
  }

  /**
   * Holds if this root directory of the application contains the given file.
   */
  predicate contains(File file) { exists(this.getFilePathRelativeToRoot(file)) }
}
