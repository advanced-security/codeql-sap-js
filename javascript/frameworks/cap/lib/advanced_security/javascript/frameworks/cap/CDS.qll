import javascript
import DataFlow

module CDS {
  // TODO: should this base type be more specific?
  abstract class ServiceInstance extends DataFlow::Node { }

  /**
   * Call to`cds.serve`
   */
  class CdsServeCall extends ServiceInstance {
    CdsServeCall() { this = any(CdsFacade cds).getMember("serve").getACall() }
  }

  /**
   * call to:
   *  `new cds.ApplicationService` or `new cds.Service`
   */
  class ServiceConstructor extends ServiceInstance {
    ServiceConstructor() { this = any(ApplicationService cds).getAnInstantiation() }
  }

  /**
   * return value of `cds.connect.to`
   */
  class ConnectTo extends ServiceInstance {
    ConnectTo() { this = any(CdsFacade cds).getMember("connect").getMember("to").getACall() }
  }

  /** Last argument to the service methods `srv.before`, `srv.on`, and `srv.after` */
  private class RequestHandler extends FunctionNode { }

  private class ErrorHandler extends RequestHandler { }

  /**
   * Subclassing ApplicationService via `extends`:
   * ```js
   * class SomeService extends cds.ApplicationService
   * ```
   */
  class UserDefinedApplicationService extends ClassNode {
    UserDefinedApplicationService() {
      exists(ApplicationService cdsApplicationService |
        this.getASuperClassNode() = cdsApplicationService.asSource()
      )
    }
  }

  /**
   * Subclassing ApplicationService via `cds.service.impl`:
   * ```js
   * const cds = require('@sap/cds')
   * module.exports = cds.service.impl (function() { ... })
   * ```
   */
  class OldStyleUserDefinedApplicationService extends MethodCallNode {
    OldStyleUserDefinedApplicationService() {
      exists(CdsFacade cds | this = cds.getMember("service").getMember("impl").getACall())
    }
  }

  /**
   *  Parameter of a `srv.with` method call:
   * ```js
   * cds.serve('./srv/cat-service') .with ((srv) => {
   *     srv.on ('READ','Books', (req) => req.reply([...]))
   * })
   * ```
   *
   * TODO expand this to capture request handlers registered inside the function
   */
  class WithCallParameter extends RequestHandler {
    WithCallParameter() {
      exists(MethodCallNode withCall, ServiceInstance svc |
        withCall.getArgument(0) = this and
        withCall.getMethodName() = "with" and
        withCall.getReceiver() = svc
      )
    }
  }

  /**
   * Parameter of request handler phases `_.before`, `_.on` `_.after`:
   * ```js
   * _.on ('READ','Books', (req) => req.reply([...]))
   * ```
   */
  class EventPhaseNodeParam extends ValueNode, ParameterNode {
    MethodCallNode eventPhase;

    EventPhaseNodeParam() {
      exists(FunctionNode handler |
        eventPhase.getMethodName() = ["before", "on", "after"] and
        eventPhase.getLastArgument() = handler and
        handler.getLastParameter() = this
      )
    }

    MethodCallNode getEventPhaseNode() { result = eventPhase }
  }

  /**
   * Parameter of request handler of `srv.on`:
   * ```js
   * this.on ('READ','Books', (req) => req.reply([...]))
   * ```
   * not sure how else to know which service is registering the handler
   */
  class RequestSource extends EventPhaseNodeParam {
    RequestSource() {
      // TODO : consider  - do we need to actually ever know which service the handler is associated to?
      exists(UserDefinedApplicationService svc, FunctionNode init |
        svc.getAnInstanceMember() = init and
        init.getName() = "init" and
        this.getEventPhaseNode().getEnclosingFunction() = init.getAstNode()
      )
      or
      exists(WithCallParameter pa |
        this.getEventPhaseNode().getEnclosingFunction() = pa.getFunction()
      )
    }
  }

  class ApplicationService extends API::Node {
    ApplicationService() { exists(CdsFacade c | this = c.getMember("ApplicationService")) }
  }

  /**
   * ```js
   * const cds = require('@sap/cds')
   * ```
   */
  class CdsFacade extends API::Node {
    CdsFacade() { this = API::moduleImport("@sap/cds") }
  }

  /**
   * Call to`cds.log`
   */
  class CdsLogCall extends API::Node {
    CdsLogCall() { this = any(CdsFacade cds).getMember("log") }
  }

  /**
   * Arguments of calls to `cds.log.{trace, debug, info, log, warn, error}`
   */
  class CdsLogSink extends DataFlow::Node {
    CdsLogSink() {
      this =
        any(CdsLogCall cdsLog)
            .getACall()
            .getAChainedMethodCall(["trace", "debug", "info", "log", "warn", "error"])
            .getAnArgument()
    }
  }

  /**
   * Methods that parse source strings into a CQL expression
   */
  class ParseSink extends DataFlow::Node {
    ParseSink() {
      this =
        any(CdsFacade cds)
            .getMember("parse")
            .getMember(["expr", "ref", "xpr"])
            .getACall()
            .getAnArgument()
    }
  }
}
