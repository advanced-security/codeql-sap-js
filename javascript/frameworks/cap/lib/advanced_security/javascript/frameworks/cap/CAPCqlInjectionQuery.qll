import javascript
import semmle.javascript.security.dataflow.SqlInjectionCustomizations
import advanced_security.javascript.frameworks.cap.CQL
import advanced_security.javascript.frameworks.cap.RemoteFlowSources
import advanced_security.javascript.frameworks.cap.dataflow.FlowSteps

class CqlClauseWithStringConcatParameter instanceof CqlClause {
  CqlClauseWithStringConcatParameter() {
    exists(DataFlow::Node queryParameter |
      (
        if this instanceof CqlInsertClause or this instanceof CqlUpsertClause
        then
          queryParameter = this.getArgument().flow() or
          queryParameter = this.getArgument().flow().(SourceNode).getAPropertyWrite().getRhs()
        else queryParameter = this.getArgument().flow()
      ) and
      exists(StringConcatenation::getAnOperand(queryParameter))
    )
  }

  Location getLocation() { result = super.getLocation() }

  string toString() { result = super.toString() }
}

class CqlShortcutMethodCallWithStringConcat instanceof CqlShortcutMethodCall {
  CqlShortcutMethodCallWithStringConcat() {
    exists(StringConcatenation::getAnOperand(super.getAQueryParameter()))
  }

  Location getLocation() { result = super.getLocation() }

  string toString() { result = super.toString() }
}

class CqlClauseParserCallWithStringConcat instanceof CqlClauseParserCall {
  CqlClauseParserCallWithStringConcat() {
    exists(StringConcatenation::getAnOperand(super.getCdlString()))
  }

  Location getLocation() { result = super.getLocation() }

  string toString() { result = super.toString() }
}

class CqlInjectionConfiguration extends TaintTracking::Configuration {
  CqlInjectionConfiguration() { this = "CQL injection from untrusted data" }

  override predicate isSource(DataFlow::Node source) { source instanceof RemoteFlowSource }

  override predicate isSink(DataFlow::Node node) {
    exists(CqlRunMethodCall cqlRunMethodCall |
      node = cqlRunMethodCall.(CqlRunMethodCall).getAQueryParameter()
    )
    or
    exists(CqlShortcutMethodCallWithStringConcat queryRunnerCall |
      node = queryRunnerCall.(CqlQueryRunnerCall).getAQueryParameter()
    )
    or
    exists(AwaitExpr await, CqlClauseWithStringConcatParameter cqlClauseWithStringConcat |
      node = await.flow() and
      await.getOperand() = cqlClauseWithStringConcat.(CqlClause).asExpr()
    )
  }

  override predicate isSanitizer(DataFlow::Node node) { node instanceof SqlInjection::Sanitizer }

  override predicate isAdditionalTaintStep(DataFlow::Node start, DataFlow::Node end) {
    exists(CqlClauseParserCallWithStringConcat cqlParseCallWithStringConcat |
      start = cqlParseCallWithStringConcat.(CqlClauseParserCall).getAnArgument() and
      end = cqlParseCallWithStringConcat
    )
    or
    exists(CqlClause cqlClause |
      start = cqlClause.getArgument().flow() and
      end = cqlClause.flow()
    )
  }
}
