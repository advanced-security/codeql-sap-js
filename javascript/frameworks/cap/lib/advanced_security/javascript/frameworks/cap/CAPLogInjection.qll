import javascript
import DataFlow
import advanced_security.javascript.frameworks.cap.CDS

/**
 * A logger obtained by a call to `log` on a CDS facade. Each logger is associated with
 * its unique name.
 */
class CdsLogger extends MethodCallNode {
  string name;

  CdsLogger() {
    this = any(CdsFacade cds).getMember("log").getACall() and
    name = this.getArgument(0).getALocalSource().asExpr().(StringLiteral).getValue()
  }
}

/**
 * Arguments of calls to `cds.log.{trace, debug, info, log, warn, error}`
 */
class CdsLogSink extends DataFlow::Node {
  CdsLogSink() {
    exists(CdsLogger log, MethodCallNode loggingMethod |
      this = loggingMethod.getAnArgument() and
      not this.asExpr() instanceof Literal and
      not this.asExpr() instanceof TemplateLiteral and
      loggingMethod.getReceiver().getALocalSource() = log and
      loggingMethod.getMethodName() = ["trace", "debug", "info", "log", "warn", "error"]
    )
  }
}
