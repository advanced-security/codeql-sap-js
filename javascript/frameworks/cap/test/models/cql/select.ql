import javascript
import advanced_security.javascript.frameworks.cap.CQL

from CqlSelectExpr s
select s.getLocation(), s