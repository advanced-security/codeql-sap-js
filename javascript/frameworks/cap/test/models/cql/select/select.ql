import javascript
import advanced_security.javascript.frameworks.cap.CQL
import advanced_security.javascript.frameworks.cap.CDS

// from CQL::CqlSelectExpr s
// select s.getLocation(), s
from API::Node cds, PropRef cdsDotQL, Node v , VarRef var
where
cds = API::moduleImport("@sap/cds") and
      cds.getMember("ql").asSource() = cdsDotQL
      and cdsDotQL.(DataFlow::SourceNode).flowsTo+(v)
      //and var.getADeclaration().getFirstControlFlowNode() = v.getAstNode().getAChild().getFirstControlFlowNode()
      and var.getName() = "SELECT"
      and v.getAstNode().getAChild() = var
      select v, 
      v.getAstNode().getAChild(),
      //v.getAstNode().getAChild().getFirstControlFlowNode(), 
     // v.getAstNode().getAChild().getAQlClass(), 
      var, var.getVariable()//.getADeclaration().getAQlClass(), var.getADeclaration().getFirstControlFlowNode()

// from Variable v 
// where v.getName() = "SELECT"
// select v, v.getADeclaration(), v.getADeclaration().getAQlClass()