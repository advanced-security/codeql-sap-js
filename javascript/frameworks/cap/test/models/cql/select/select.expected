| select2.js:4:14:4:35 | select2.js:4 | select2.js:4:14:4:35 | SELECT. ... (Table) |
| select.js:3:14:3:43 | select.js:3 | select.js:3:14:3:43 | SELECT` ... (Table) |
| select.js:4:14:4:45 | select.js:4 | select.js:4:14:4:45 | SELECT` ... Table") |
| select.js:5:14:5:33 | select.js:5 | select.js:5:14:5:33 | SELECT("col1, col2") |
| select.js:6:14:6:33 | select.js:6 | select.js:6:14:6:33 | SELECT("col1, col2") |
| select.js:6:14:6:45 | select.js:6 | select.js:6:14:6:45 | SELECT( ... (Table) |
| select.js:7:14:7:33 | select.js:7 | select.js:7:14:7:33 | SELECT("col1, col2") |
| select.js:7:14:7:47 | select.js:7 | select.js:7:14:7:47 | SELECT( ... Table") |
| select.js:11:14:11:35 | select.js:11 | select.js:11:14:11:35 | SELECT. ... (Table) |
| select.js:13:14:13:40 | select.js:13 | select.js:13:14:13:40 | SELECT. ... (Table) |
| select.js:14:14:14:54 | select.js:14 | select.js:14:14:14:54 | SELECT. ... (Table) |
| select.js:20:14:22:2 | select.js:20 | select.js:20:14:22:2 | SELECT. ... ol2;\\n}) |
| select.js:24:14:24:68 | select.js:24 | select.js:24:14:24:68 | SELECT. ... Alias") |
| select.js:25:14:28:2 | select.js:25 | select.js:25:14:28:2 | SELECT. ... " },\\n]) |
| select.js:29:14:32:2 | select.js:29 | select.js:29:14:32:2 | SELECT. ... ty",\\n}) |
| select.js:34:14:34:31 | select.js:34 | select.js:34:14:34:31 | SELECT.from(Table) |
| select.js:35:14:35:31 | select.js:35 | select.js:35:14:35:31 | SELECT.from(Table) |
| select.js:35:14:37:2 | select.js:35 | select.js:35:14:37:2 | SELECT. ... ol2;\\n}) |
| select.js:38:14:38:31 | select.js:38 | select.js:38:14:38:31 | SELECT.from(Table) |
| select.js:39:14:39:31 | select.js:39 | select.js:39:14:39:31 | SELECT.from(Table) |
| select.js:39:14:39:68 | select.js:39 | select.js:39:14:39:68 | SELECT. ... Alias") |
| select.js:40:14:40:31 | select.js:40 | select.js:40:14:40:31 | SELECT.from(Table) |
| select.js:40:14:43:2 | select.js:40 | select.js:40:14:43:2 | SELECT. ... " },\\n]) |
| select.js:44:14:44:31 | select.js:44 | select.js:44:14:44:31 | SELECT.from(Table) |
| select.js:44:14:47:2 | select.js:44 | select.js:44:14:47:2 | SELECT. ... ty",\\n}) |
| select.js:50:14:50:52 | select.js:50 | select.js:50:14:50:52 | SELECT. ...  "*" }) |
| select.js:51:14:51:49 | select.js:51 | select.js:51:14:51:49 | SELECT. ... 1='*'") |
| select.js:53:14:53:51 | select.js:53 | select.js:53:14:53:51 | SELECT. ... ", "*") |
| select.js:55:14:55:58 | select.js:55 | select.js:55:14:55:58 | SELECT. ... , 10)") |
| select.js:57:14:57:70 | select.js:57 | select.js:57:14:57:70 | SELECT. ... 11 } }) |
| select.js:59:14:59:31 | select.js:59 | select.js:59:14:59:31 | SELECT.from(Table) |
| select.js:59:14:59:52 | select.js:59 | select.js:59:14:59:52 | SELECT. ...  "*" }) |
| select.js:60:14:60:31 | select.js:60 | select.js:60:14:60:31 | SELECT.from(Table) |
| select.js:60:14:60:49 | select.js:60 | select.js:60:14:60:49 | SELECT. ... 1='*'") |
| select.js:61:14:61:31 | select.js:61 | select.js:61:14:61:31 | SELECT.from(Table) |
| select.js:62:14:62:31 | select.js:62 | select.js:62:14:62:31 | SELECT.from(Table) |
| select.js:62:14:62:51 | select.js:62 | select.js:62:14:62:51 | SELECT. ... ", "*") |
| select.js:63:14:63:31 | select.js:63 | select.js:63:14:63:31 | SELECT.from(Table) |
| select.js:64:14:64:31 | select.js:64 | select.js:64:14:64:31 | SELECT.from(Table) |
| select.js:64:14:64:58 | select.js:64 | select.js:64:14:64:58 | SELECT. ... , 10)") |
| select.js:65:14:65:31 | select.js:65 | select.js:65:14:65:31 | SELECT.from(Table) |
| select.js:66:14:66:31 | select.js:66 | select.js:66:14:66:31 | SELECT.from(Table) |
| select.js:66:14:66:70 | select.js:66 | select.js:66:14:66:70 | SELECT. ... 11 } }) |
| select.js:69:14:69:55 | select.js:69 | select.js:69:14:69:55 | SELECT. ... "col2") |
| select.js:71:14:71:67 | select.js:71 | select.js:71:14:71:67 | SELECT. ... prop2") |
| select.js:73:14:76:1 | select.js:73 | select.js:73:14:76:1 | SELECT. ... 2"] }\\n) |
| select.js:78:14:78:31 | select.js:78 | select.js:78:14:78:31 | SELECT.from(Table) |
| select.js:78:14:78:55 | select.js:78 | select.js:78:14:78:55 | SELECT. ... "col2") |
| select.js:79:14:79:31 | select.js:79 | select.js:79:14:79:31 | SELECT.from(Table) |
| select.js:80:14:80:31 | select.js:80 | select.js:80:14:80:31 | SELECT.from(Table) |
| select.js:80:14:80:67 | select.js:80 | select.js:80:14:80:67 | SELECT. ... prop2") |
| select.js:81:14:81:31 | select.js:81 | select.js:81:14:81:31 | SELECT.from(Table) |
| select.js:82:14:82:31 | select.js:82 | select.js:82:14:82:31 | SELECT.from(Table) |
| select.js:82:14:85:1 | select.js:82 | select.js:82:14:85:1 | SELECT. ... 2"] }\\n) |
| select.js:88:14:88:53 | select.js:88 | select.js:88:14:88:53 | SELECT. ...  "*" }) |
| select.js:89:14:89:50 | select.js:89 | select.js:89:14:89:50 | SELECT. ... 1='*'") |
| select.js:91:14:91:52 | select.js:91 | select.js:91:14:91:52 | SELECT. ... ", "*") |
| select.js:93:14:93:59 | select.js:93 | select.js:93:14:93:59 | SELECT. ... , 10)") |
| select.js:95:14:95:71 | select.js:95 | select.js:95:14:95:71 | SELECT. ... 11 } }) |
| select.js:97:14:97:31 | select.js:97 | select.js:97:14:97:31 | SELECT.from(Table) |
| select.js:97:14:97:53 | select.js:97 | select.js:97:14:97:53 | SELECT. ...  "*" }) |
| select.js:98:14:98:31 | select.js:98 | select.js:98:14:98:31 | SELECT.from(Table) |
| select.js:98:14:98:50 | select.js:98 | select.js:98:14:98:50 | SELECT. ... 1='*'") |
| select.js:99:14:99:31 | select.js:99 | select.js:99:14:99:31 | SELECT.from(Table) |
| select.js:100:14:100:31 | select.js:100 | select.js:100:14:100:31 | SELECT.from(Table) |
| select.js:100:14:100:52 | select.js:100 | select.js:100:14:100:52 | SELECT. ... ", "*") |
| select.js:101:14:101:31 | select.js:101 | select.js:101:14:101:31 | SELECT.from(Table) |
| select.js:102:14:102:31 | select.js:102 | select.js:102:14:102:31 | SELECT.from(Table) |
| select.js:102:14:102:59 | select.js:102 | select.js:102:14:102:59 | SELECT. ... , 10)") |
| select.js:103:14:103:31 | select.js:103 | select.js:103:14:103:31 | SELECT.from(Table) |
| select.js:104:14:104:31 | select.js:104 | select.js:104:14:104:31 | SELECT.from(Table) |
| select.js:104:14:104:71 | select.js:104 | select.js:104:14:104:71 | SELECT. ... 11 } }) |
| select.js:106:14:106:55 | select.js:106 | select.js:106:14:106:55 | SELECT. ... "col2") |
| select.js:106:14:106:77 | select.js:106 | select.js:106:14:106:77 | SELECT. ...  "*" }) |
| select.js:107:14:107:55 | select.js:107 | select.js:107:14:107:55 | SELECT. ... "col2") |
| select.js:107:14:107:74 | select.js:107 | select.js:107:14:107:74 | SELECT. ... 1='*'") |
| select.js:108:14:108:55 | select.js:108 | select.js:108:14:108:55 | SELECT. ... "col2") |
| select.js:109:14:109:55 | select.js:109 | select.js:109:14:109:55 | SELECT. ... "col2") |
| select.js:109:14:109:76 | select.js:109 | select.js:109:14:109:76 | SELECT. ... ", "*") |
| select.js:110:14:110:55 | select.js:110 | select.js:110:14:110:55 | SELECT. ... "col2") |
| select.js:111:14:112:26 | select.js:111 | select.js:111:14:112:26 | SELECT. ... "col2") |
| select.js:111:14:113:30 | select.js:111 | select.js:111:14:113:30 | SELECT. ... , 10)") |
| select.js:114:14:114:55 | select.js:114 | select.js:114:14:114:55 | SELECT. ... "col2") |
| select.js:117:14:118:26 | select.js:117 | select.js:117:14:118:26 | SELECT. ... "col2") |
| select.js:117:14:119:42 | select.js:117 | select.js:117:14:119:42 | SELECT. ... 11 } }) |
| select.js:121:14:121:31 | select.js:121 | select.js:121:14:121:31 | SELECT.from(Table) |
| select.js:121:14:121:55 | select.js:121 | select.js:121:14:121:55 | SELECT. ... "col2") |
| select.js:121:14:121:77 | select.js:121 | select.js:121:14:121:77 | SELECT. ...  "*" }) |
| select.js:122:14:122:31 | select.js:122 | select.js:122:14:122:31 | SELECT.from(Table) |
| select.js:122:14:122:55 | select.js:122 | select.js:122:14:122:55 | SELECT. ... "col2") |
| select.js:122:14:122:74 | select.js:122 | select.js:122:14:122:74 | SELECT. ... 1='*'") |
| select.js:123:14:123:31 | select.js:123 | select.js:123:14:123:31 | SELECT.from(Table) |
| select.js:123:14:123:55 | select.js:123 | select.js:123:14:123:55 | SELECT. ... "col2") |
| select.js:124:14:124:31 | select.js:124 | select.js:124:14:124:31 | SELECT.from(Table) |
| select.js:124:14:124:55 | select.js:124 | select.js:124:14:124:55 | SELECT. ... "col2") |
| select.js:124:14:124:76 | select.js:124 | select.js:124:14:124:76 | SELECT. ... ", "*") |
| select.js:125:14:125:31 | select.js:125 | select.js:125:14:125:31 | SELECT.from(Table) |
| select.js:125:14:125:55 | select.js:125 | select.js:125:14:125:55 | SELECT. ... "col2") |
| select.js:126:14:126:31 | select.js:126 | select.js:126:14:126:31 | SELECT.from(Table) |
| select.js:126:14:127:26 | select.js:126 | select.js:126:14:127:26 | SELECT. ... "col2") |
| select.js:126:14:128:30 | select.js:126 | select.js:126:14:128:30 | SELECT. ... , 10)") |
| select.js:129:14:129:31 | select.js:129 | select.js:129:14:129:31 | SELECT.from(Table) |
| select.js:129:14:129:55 | select.js:129 | select.js:129:14:129:55 | SELECT. ... "col2") |
| select.js:132:14:132:31 | select.js:132 | select.js:132:14:132:31 | SELECT.from(Table) |
| select.js:132:14:133:26 | select.js:132 | select.js:132:14:133:26 | SELECT. ... "col2") |
| select.js:132:14:134:42 | select.js:132 | select.js:132:14:134:42 | SELECT. ... 11 } }) |
| select.js:136:14:136:73 | select.js:136 | select.js:136:14:136:73 | SELECT. ...  "*" }) |
| select.js:137:14:137:70 | select.js:137 | select.js:137:14:137:70 | SELECT. ... 1='*'") |
| select.js:139:14:139:72 | select.js:139 | select.js:139:14:139:72 | SELECT. ... ", "*") |
| select.js:141:14:141:79 | select.js:141 | select.js:141:14:141:79 | SELECT. ... , 10)") |
| select.js:145:14:148:2 | select.js:145 | select.js:145:14:148:2 | SELECT. ... 1 },\\n}) |
| select.js:150:14:150:51 | select.js:150 | select.js:150:14:150:51 | SELECT. ... , col2) |
| select.js:150:14:150:73 | select.js:150 | select.js:150:14:150:73 | SELECT. ...  "*" }) |
| select.js:151:14:151:51 | select.js:151 | select.js:151:14:151:51 | SELECT. ... , col2) |
| select.js:151:14:151:70 | select.js:151 | select.js:151:14:151:70 | SELECT. ... 1='*'") |
| select.js:152:14:152:51 | select.js:152 | select.js:152:14:152:51 | SELECT. ... , col2) |
| select.js:153:14:153:51 | select.js:153 | select.js:153:14:153:51 | SELECT. ... , col2) |
| select.js:153:14:153:72 | select.js:153 | select.js:153:14:153:72 | SELECT. ... ", "*") |
| select.js:154:14:154:51 | select.js:154 | select.js:154:14:154:51 | SELECT. ... , col2) |
| select.js:155:14:155:51 | select.js:155 | select.js:155:14:155:51 | SELECT. ... , col2) |
| select.js:155:14:155:79 | select.js:155 | select.js:155:14:155:79 | SELECT. ... , 10)") |
| select.js:156:14:156:51 | select.js:156 | select.js:156:14:156:51 | SELECT. ... , col2) |
| select.js:159:14:159:51 | select.js:159 | select.js:159:14:159:51 | SELECT. ... , col2) |
| select.js:159:14:162:2 | select.js:159 | select.js:159:14:162:2 | SELECT. ... 1 },\\n}) |
| select.js:164:14:165:38 | select.js:164 | select.js:164:14:165:38 | SELECT. ... prop2") |
| select.js:164:14:166:24 | select.js:164 | select.js:164:14:166:24 | SELECT. ...  "*" }) |
| select.js:167:14:168:38 | select.js:167 | select.js:167:14:168:38 | SELECT. ... prop2") |
| select.js:167:14:169:21 | select.js:167 | select.js:167:14:169:21 | SELECT. ... 1='*'") |
| select.js:170:14:170:67 | select.js:170 | select.js:170:14:170:67 | SELECT. ... prop2") |
| select.js:172:14:173:38 | select.js:172 | select.js:172:14:173:38 | SELECT. ... prop2") |
| select.js:172:14:174:23 | select.js:172 | select.js:172:14:174:23 | SELECT. ... ", "*") |
| select.js:175:14:175:67 | select.js:175 | select.js:175:14:175:67 | SELECT. ... prop2") |
| select.js:177:14:178:38 | select.js:177 | select.js:177:14:178:38 | SELECT. ... prop2") |
| select.js:177:14:179:30 | select.js:177 | select.js:177:14:179:30 | SELECT. ... , 10)") |
| select.js:180:14:180:67 | select.js:180 | select.js:180:14:180:67 | SELECT. ... prop2") |
| select.js:182:14:183:38 | select.js:182 | select.js:182:14:183:38 | SELECT. ... prop2") |
| select.js:182:14:184:42 | select.js:182 | select.js:182:14:184:42 | SELECT. ... 11 } }) |
| select.js:186:14:188:2 | select.js:186 | select.js:186:14:188:2 | SELECT. ... "*",\\n}) |
| select.js:189:14:191:1 | select.js:189 | select.js:189:14:191:1 | SELECT. ... ='*'"\\n) |
| select.js:194:14:197:1 | select.js:194 | select.js:194:14:197:1 | SELECT. ...   "*"\\n) |
| select.js:200:14:202:1 | select.js:200 | select.js:200:14:202:1 | SELECT. ...  10)"\\n) |
| select.js:205:14:208:2 | select.js:205 | select.js:205:14:208:2 | SELECT. ... 1 },\\n}) |
| select.js:210:14:211:66 | select.js:210 | select.js:210:14:211:66 | SELECT. ... p2"] }) |
| select.js:210:14:212:24 | select.js:210 | select.js:210:14:212:24 | SELECT. ...  "*" }) |
| select.js:213:14:214:66 | select.js:213 | select.js:213:14:214:66 | SELECT. ... p2"] }) |
| select.js:213:14:215:21 | select.js:213 | select.js:213:14:215:21 | SELECT. ... 1='*'") |
| select.js:216:14:219:1 | select.js:216 | select.js:216:14:219:1 | SELECT. ... 2"] }\\n) |
| select.js:220:14:221:66 | select.js:220 | select.js:220:14:221:66 | SELECT. ... p2"] }) |
| select.js:220:14:222:23 | select.js:220 | select.js:220:14:222:23 | SELECT. ... ", "*") |
| select.js:223:14:226:1 | select.js:223 | select.js:223:14:226:1 | SELECT. ... 2"] }\\n) |
| select.js:227:14:228:66 | select.js:227 | select.js:227:14:228:66 | SELECT. ... p2"] }) |
| select.js:227:14:229:30 | select.js:227 | select.js:227:14:229:30 | SELECT. ... , 10)") |
| select.js:230:14:233:1 | select.js:230 | select.js:230:14:233:1 | SELECT. ... 2"] }\\n) |
| select.js:234:14:235:66 | select.js:234 | select.js:234:14:235:66 | SELECT. ... p2"] }) |
| select.js:234:14:236:42 | select.js:234 | select.js:234:14:236:42 | SELECT. ... 11 } }) |
| select.js:248:14:248:31 | select.js:248 | select.js:248:14:248:31 | SELECT.from(Table) |
| select.js:249:14:249:31 | select.js:249 | select.js:249:14:249:31 | SELECT.from(Table) |
| select.js:250:14:250:31 | select.js:250 | select.js:250:14:250:31 | SELECT.from(Table) |
| select.js:251:14:251:31 | select.js:251 | select.js:251:14:251:31 | SELECT.from(Table) |
| select.js:252:14:252:31 | select.js:252 | select.js:252:14:252:31 | SELECT.from(Table) |
| select.js:253:14:253:31 | select.js:253 | select.js:253:14:253:31 | SELECT.from(Table) |
| select.js:254:14:254:31 | select.js:254 | select.js:254:14:254:31 | SELECT.from(Table) |
| select.js:255:14:255:31 | select.js:255 | select.js:255:14:255:31 | SELECT.from(Table) |
| select.js:258:14:258:41 | select.js:258 | select.js:258:14:258:41 | SELECT. ... mit(10) |
| select.js:260:14:260:45 | select.js:260 | select.js:260:14:260:45 | SELECT. ... 10, 20) |
| select.js:261:14:261:50 | select.js:261 | select.js:261:14:261:50 | SELECT. ... : 10 }) |
| select.js:262:14:262:63 | select.js:262 | select.js:262:14:262:63 | SELECT. ... : 20 }) |
| select.js:263:14:263:60 | select.js:263 | select.js:263:14:263:60 | SELECT. ... al"] }) |
| select.js:264:14:266:2 | select.js:264 | select.js:264:14:266:2 | SELECT. ...  }],\\n}) |
| select.js:268:14:268:31 | select.js:268 | select.js:268:14:268:31 | SELECT.from(Table) |
| select.js:268:14:268:41 | select.js:268 | select.js:268:14:268:41 | SELECT. ... mit(10) |
| select.js:269:14:269:31 | select.js:269 | select.js:269:14:269:31 | SELECT.from(Table) |
| select.js:270:14:270:31 | select.js:270 | select.js:270:14:270:31 | SELECT.from(Table) |
| select.js:270:14:270:45 | select.js:270 | select.js:270:14:270:45 | SELECT. ... 10, 20) |
| select.js:271:14:271:31 | select.js:271 | select.js:271:14:271:31 | SELECT.from(Table) |
| select.js:271:14:271:50 | select.js:271 | select.js:271:14:271:50 | SELECT. ... : 10 }) |
| select.js:272:14:272:31 | select.js:272 | select.js:272:14:272:31 | SELECT.from(Table) |
| select.js:272:14:272:63 | select.js:272 | select.js:272:14:272:63 | SELECT. ... : 20 }) |
| select.js:273:14:273:31 | select.js:273 | select.js:273:14:273:31 | SELECT.from(Table) |
| select.js:273:14:273:60 | select.js:273 | select.js:273:14:273:60 | SELECT. ... al"] }) |
| select.js:274:14:274:31 | select.js:274 | select.js:274:14:274:31 | SELECT.from(Table) |
| select.js:274:14:276:2 | select.js:274 | select.js:274:14:276:2 | SELECT. ...  }],\\n}) |
| select.js:279:14:280:35 | select.js:279 | select.js:279:14:280:35 | SELECT. ... pdate() |
| select.js:283:14:284:38 | select.js:283 | select.js:283:14:284:38 | SELECT. ... eLock() |
| select.js:287:14:290:1 | select.js:287 | select.js:287:14:290:1 | SELECT. ... col2"\\n) |
| select.js:287:14:292:2 | select.js:287 | select.js:287:14:292:2 | SELECT. ...  }],\\n}) |
| select.js:287:14:292:48 | select.js:287 | select.js:287:14:292:48 | SELECT. ... eLock() |
| select.js:325:10:325:28 | select.js:325 | select.js:325:10:325:28 | SELECT.from (Books) |
